File: cleaner/cleaner.go
// Package cleaner provides a type that can remove all node.Node's from a
// filesystem.
package cleaner

import (
	"github.com/ankeesler/btool/log"
	"github.com/ankeesler/btool/node"
	"github.com/pkg/errors"
)

//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . RemoveAller

// RemoveAller is a type that can delete a file from the filesystem. If the file
// does not exist, then it shouldn't fail.
type RemoveAller interface {
	RemoveAll(string) error
}

//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 . Callback

// Callback is an interface that clients can use to be notified when a node.Node
// is removed from the filesystem.
type Callback interface {
	OnClean(*node.Node)
}

// Cleaner can remove all node.Node's from a filesystem.
type Cleaner struct {
	ra RemoveAller
	cb Callback
}

// New creates a new Cleaner.
func New(ra RemoveAller, cb Callback) *Cleaner {
	return &Cleaner{
		ra: ra,
		cb: cb,
	}
}

// Clean will walk a node.Node graph and remove all node.Node's from the
// filesystem.
func (c *Cleaner) Clean(n *node.Node) error {
	return c.clean(n, make(map[*node.Node]bool))
}

func (c *Cleaner) clean(n *node.Node, cleaned map[*node.Node]bool) error {
	if cleaned[n] {
		return nil
	}

	log.Debugf("cleaning %s", n.Name)

	for _, dN := range n.Dependencies {
		if err := c.clean(dN, cleaned); err != nil {
			return errors.Wrap(err, "clean")
		}
	}

	if n.Resolver != nil {
		c.cb.OnClean(n)
		if err := c.ra.RemoveAll(n.Name); err != nil {
			return errors.Wrap(err, "remove all")
		}
	}

	cleaned[n] = true

	return nil
}

File: cc/app/cleaner/cleaner.cc
#include "app/cleaner/cleaner.h"

#include <string>

#include "err.h"
#include "log.h"
#include "node/node.h"

namespace btool::app::cleaner {

bool Cleaner::Clean(const ::btool::node::Node& node, std::string* ret_err) {
  bool success = true;

  node.Visit([&](const ::btool::node::Node* n) {
    if (success && n->resolver() != nullptr) {
      DEBUGS() << "cleaning " << n->name() << std::endl;

      std::string err;
      if (!ra_->RemoveAll(n->name(), &err)) {
        *ret_err = ::btool::WrapErr(err, "remove all");
        success = false;
      }
    }
  });

  return success;
}

};  // namespace btool::app::cleaner
